src/state.rs

```diff
@@ -0,0 +1,415 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::identity::SecretManager;
+use crate::metrics::Metrics;
+use crate::proxy::Error;
+use crate::state::service::ServiceStore;
+use crate::state::workload::address::Address;
+use crate::state::workload::{
+    gatewayaddress, network_addr, NetworkAddress, Protocol, WaypointError, Workload, WorkloadStore,
+};
+use crate::xds::{AdsClient, Demander, LocalClient, ProxyStateUpdater};
+use crate::{cert_fetcher, config, rbac, readiness, xds};
+use rand::prelude::IteratorRandom;
+use rand::seq::SliceRandom;
+use std::convert::Into;
+use std::default::Default;
+use std::fmt;
+use std::net::{IpAddr, SocketAddr};
+use std::sync::{Arc, RwLock};
+use tracing::{debug, trace};
+
+pub mod service;
+pub mod workload;
+
+#[derive(Debug, Hash, Eq, PartialEq, Clone, serde::Serialize)]
+pub struct Upstream {
+    pub workload: Workload,
+    pub port: u16,
+}
+
+impl fmt::Display for Upstream {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Upstream{{{} with uid {}:{} via {} ({:?})}}",
+            self.workload.name,
+            self.workload.uid,
+            self.port,
+            self.workload
+                .gateway_address
+                .map(|x| format!("{x}"))
+                .unwrap_or_else(|| "None".into()),
+            self.workload.protocol
+        )
+    }
+}
+
+/// The current state information for this proxy.
+#[derive(serde::Serialize, Default, Debug)]
+pub struct ProxyState {
+    #[serde(flatten)]
+    pub workloads: WorkloadStore,
+
+    #[serde(flatten)]
+    pub services: ServiceStore,
+}
+
+/// Wrapper around [ProxyState] that provides additional methods for requesting information
+/// on-demand.
+#[derive(serde::Serialize, Debug, Clone)]
+pub struct DemandProxyState {
+    #[serde(flatten)]
+    pub state: Arc<RwLock<ProxyState>>,
+
+    /// If present, used to request on-demand updates for workloads.
+    #[serde(skip_serializing)]
+    demand: Option<Demander>,
+}
+
+impl DemandProxyState {
```

Comment:
Note to reviewers: The functions here are just a direct copy of those from `WorkloadInformation`.